### PR TITLE
rgw_sal_motr: [CORTX-31877] Support Stat counting for Multipart & Obj OverWrite case

### PR DESCRIPTION
**Pull request by sumedhak27**
In multipart upload, each part upload increments the total object size by part size
and the count by one. When the complete multipart is called, the object count is reset to
consider all parts as a single object.
On abort size of each part is subtracted and the object count is decremented by the number
of parts as well.

Object overwrite case is also handled with the same PR which decreases the size of the old
obj and increments the size of the new obj in case of unversioned bucket. Obj count will
remain the same.

Signed-off-by: Sumedh A. Kulkarni <sumedh.a.kulkarni@seagate.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/rgw/rgw_sal_motr.cc [hunk 1]
```diff
@@ -176,6 +176,35 @@ inline std::string get_bucket_name(const std::string& tenant,  const std::string
     return bucket;
 }
 
+int static update_bucket_stats(const DoutPrefixProvider *dpp, MotrStore *store,
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:179]: Any reasons to keep this function static?

File: src/rgw/rgw_sal_motr.cc [hunk 2]
```diff
@@ -176,6 +176,35 @@ inline std::string get_bucket_name(const std::string& tenant,  const std::string
...
+                               std::string owner, std::string bucket_name,
+                               uint64_t size, uint64_t actual_size,
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:181]: Please add a comment stating what `size` and `actual_size` exactly represents.

File: src/rgw/rgw_sal_motr.cc [hunk 3]
```diff
@@ -2108,12 +2106,28 @@ int MotrObject::remove_mobj_and_index_entry(const DoutPrefixProvider* dpp, rgw_b
...
     return rc;
   }
+
+  // Subtract object size & count from the bucket stats.
+  if (ent.is_delete_marker())
+    return rc;
+  rc = update_bucket_stats(dpp, this->store, ent.meta.owner, bucket_name,
+                            ent.meta.size, ent.meta.size, 1, false);
+  if (rc != 0) {
+    ldpp_dout(dpp, 20) << __func__ << ": Failed stats substraction for the "
```
Comment (jjxsg) [inline src/rgw/rgw_sal_motr.cc:2122]: Change it to `Stats subtraction failed for `

**Review by jjxsg — COMMENTED** [review]
We need to evaluate this with IAM policy as well. Try to add an object in user1's bucket, give user2 rights to perform `PutObject` & `DeleteObject` actions on that bucket, try deleting/adding objects from user2 and check if stats are being updated correctly for both simple and MP objects.
@sumedhak27 @Diwakar92 please test this with both of your changes.

File: src/rgw/rgw_sal_motr.cc (hunk 3 above)
Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:2122]: sure will address with the next PR regarding Stats.

File: src/rgw/rgw_sal_motr.cc (hunk 2 above)
Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:181]: sure will address with the next PR regarding Stats.

File: src/rgw/rgw_sal_motr.cc (hunk 1 above)
Comment (sumedhak27) [reply src/rgw/rgw_sal_motr.cc:179]: sure will address with the next PR regarding Stats.

**Comment by sumedhak27** [comment]
Changes are addressed via https://github.com/Seagate/cortx-rgw/pull/292.
Closing the PR.

**Pull request closed without merging**